src/metabase/models/dashboard_card.clj

```diff
@@ -217,6 +221,117 @@
+
+(defn- link-card-columns-for-model
+  [model]
+  (let [model-cols-set                   (link-card-columns-for-model* model)
+        model-col-alias->honeysql-clause (m/index-by ->column-alias model-cols-set)]
+    (for [[col col-type] all-card-info-columns
+          :let           [maybe-aliased-col (get model-col-alias->honeysql-clause col)]]
+      (cond
+        (= col :model)
+        [(h2x/literal model) :model]
+
+        maybe-aliased-col
+        maybe-aliased-col
+
+        ;; This entity is missing the column, project a null for that column value. For Postgres and H2, cast it to the
+        ;; correct type, e.g.
+        ;;
+        ;;    SELECT cast(NULL AS integer)
+        ;;
+        ;; For MySQL, this is not needed.
+        :else
+        [(when-not (= (mdb/db-type) :mysql)
+           [:cast nil col-type])
+         col]))))
+
+(def ^:private link-card-model->toucan-model
+  {:database  :metabase.models.database/Database
+   :table     :metabase.models.table/Table
+   :dashboard :metabase.models.dashboard/Dashboard
+   :card      :metabase.models.card/Card
+   :dataset   :metabase.models.card/Card})
+
+(defn- link-card-info-query-for-model
+  [[model ids]]
+  {:select (link-card-columns-for-model model)
+   :from   (t2/table-name (link-card-model->toucan-model model))
+   :where  [:in :id ids]})
+
+(defn- link-card-info-query
+  [link-card-model->ids]
+  (if (= 1 (count link-card-model->ids))
+    (link-card-info-query-for-model (first link-card-model->ids))
+    {:select   [:*]
+     :from     [[{:union-all (map link-card-info-query-for-model link-card-model->ids)}
+                 :alias_is_required_by_sql_but_not_needed_here]]}))
+
+(def ^:private link-card-models
+  #{"database" "table" "dataset" "card" "dashboard"})
+
+(defn with-link-card-info
+  "Update entity info for link cards.
+
+  Link cards are dashcards that link to internal entities like Database/Dashboard/... or an url.
+  The viz-settings only store the model name and id, info like name, description will need to be
+  hydrated on fetch to make sure those info are up-to-date."
+  [dashcards]
+  (let [entity-path        [:visualization_settings :link :entity]
+        ;; find all dashcards that are link-cards and get its model, id
+        ;; [[:table #{1 2}] [:database #{3 4}]]
+        model-and-ids      (->> dashcards
+                                (filter #(link-card-models (get-in % (conj entity-path :model))))
+                                (map #(get-in % entity-path))
+                                (group-by :model)
+                                (map (fn [[k v]] [(keyword k) (set (map :id v))])))
+        ;; query all entities in 1 db call
+        ;; {[:table 3] {:name ...}}
+        model-and-id->info (when (seq model-and-ids)
+                             (-> (m/index-by (juxt :model :id) (t2/query (link-card-info-query model-and-ids)))
+                                 (update-vals (fn [{model :model :as instance}]
```

Comment:
things that come out of `t2/query` are instances without a model, In a sense it's appropriate to call them `instance.

But I understand your concern, maybe we need a better way to name things?

Reply:
> It's cleaner to do the index last I think
(-> (map ...) (m/index-by ...))
than
(-> (m/index-by ...) (update-vals ...)

do you mean we should mask with `{:restricted true}` before index-by?

if so we can't do that because we need `:model :id` to index them.